### PR TITLE
Improve VoiceOver integration in Menus section - Mark I

### PR DESCRIPTION
**Pull request by etoledom**
Fixes parts of #12881

This PR improves VoiceOver integration of the Header part of the Menus section.

Menus

VoiceOver reads the Areas picker cell like this:
> Menu area: Header. 3 menu areas in this theme. Button. [hint] Expands to select a different menu area.

I'm not sure about the `3 menu areas in this theme` part. It might be useful information but makes the label a bit too long.
What do you think @iamthomasbishop ?

Same with the Menu picker cell. It reads:
> Menu in area Header: Primary. 3 menus available. Button. [hint] Expands to select a different menu to edit.

### To test:

[Using Voice Over](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/docs/using-voiceover.md):
- Go to the Menu section of a site.
- Check that the header section of the controller reacts as expected to VoiceOver.
- Expand the Areas and Menus section.
- Also test switch actions (swipes to the left and right).
- Test landscape where Areas and Menus are side by side.
  - Check that the areas and menus are navigated as a group via switch.
- Check out the [guidelines](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/docs/voiceover-guidelines.md#questions-to-ask).

cc @designsimply 

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by iamthomasbishop** [comment]
> I'm not sure about the 3 menu areas in this theme part. It might be useful information but makes the label a bit too long.
What do you think @iamthomasbishop ?

@etoledom Maybe we could shorten this to "3 menu areas"? 

> Same with the Menu picker cell. It reads:
> Menu in area Header: Primary. 3 menus available. Button. [hint] Expands to select a different menu to edit.

If we want to shorten this, what about something like "Menu in Header area: 3 menus available. Button. [hint] Expands to select a different menu to edit."

File: WordPress/Classes/ViewRelated/Menus/MenusSelectionItemView.m [hunk 1]
```diff
@@ -194,6 +196,22 @@ - (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
...
+{
+    self.isAccessibilityElement = YES;
+    self.accessibilityTraits = UIAccessibilityTraitButton;
+    self.accessibilityLabel = self.item.displayName;
+
+    NSString *selectedLocalizedString = NSLocalizedString(@"Selected", @"Screen reader text to represent the selected state of a button");
+    self.accessibilityValue = self.item.selected ? selectedLocalizedString : nil;
+    if (self.item.selected) {
+        self.accessibilityHint = NSLocalizedString(@"Selects this item", @"Screen reader hint (non-imperative) about what does the menu item selection button do");
+    }
```
Comment (guarani) [inline WordPress/Classes/ViewRelated/Menus/MenusSelectionItemView.m]: In tests, VoiceOver read the "Selects this item" hint on _both_ selected and non-selected items. I thought this hint should only be added to non-selected items, to describe the result of the element's action. And since already selected items can't be further selected, I didn't expect to hear it there either.

Fwiw, I think this `prepareForVoiceOver` is being called only on item initialization, so `self.item.selected` never changes (it's `true` in all my tests, even when selecting different items).

File: WordPress/Classes/ViewRelated/Menus/MenusSelectionView.m [hunk 2]
```diff
@@ -266,4 +281,46 @@ - (void)selectionItemObjectWasUpdatedNotification:(NSNotification *)notification
...
+                                          self.detailView.subTitleLabel.text];
+    self.detailView.accessibilityHint = NSLocalizedString(@"Expands to select a different menu area", @"Screen reader hint (non-imperative) about what does the site menu area selector button do.");
+}
+
+- (void)configureMenuAccessibility
+{
+    // Menu in area Header: Primary. 3 menus available. Button. [hint] Expands to select a different menu to edit.
+    NSString *format = NSLocalizedString(@"Menu in area %@: %@, %@", @"Screen reader string too choose a menu to edit. First %@ is the name of the menu area (Primary, Footer, etc...). Second %@ is name of the menu currently selected. Third is an already localized string saying the number of menus available in the menu area selected.");
+    self.detailView.accessibilityLabel = [NSString stringWithFormat:format,
+                                          self.selectedItemLocation.displayName,
```
Comment (guarani) [inline WordPress/Classes/ViewRelated/Menus/MenusSelectionView.m]: I find the presence of `selectedItemLocation` in addition to the existing `selectedItem` a bit confusing tbh. I understand that this is used to reference the selected item _from the section above_ (a different `MenusSelectionView` instance, i.e. `MenuHeaderViewController.locationsView`), but it doesn't feel like it belongs here. Especially if it's not actually needed/used in the section above.

I'd suggest using a delegate to ask the `MenuHeaderViewController` which menu location is currently active, and use that — what do you think?

**Review by guarani — COMMENTED** [review]
I tested this and overall it works great, but I found one potential bug. Here's the steps to reproduce:
1. In the top section (menu location), select a different selected menu location (changed from "Header" to "Social Menu" in my test)
2. With VoiceOver on, tap the header view (i.e. `MenusSelectionDetailView`) once to hear its accessibility message

I heard:

> Menu in area Header: No Menu. 2 menus available. Expanded. Button. [hint] Expands to select a different menu

but I expected:

> Menu in area **Social Menu**: No Menu. 2 menus available. Expanded. Button. [hint] Expands to select a different menu

because "Social Menu", not "Header", is the currently selected menu location.

**Comment by guarani** [comment]
Regarding the accessibility messages in general, e.g.:

> Menu area: Header. 3 menu areas in this theme. Button. [hint] Expands to select a different menu area.

Would it make sense to change this like so:

> Header menu area. 3 menu areas in this theme. Button. [hint] Expands to select a different menu area.

This would feel more natural to me. Keeping in mind that "Header" could be "Social Menu" or any other name, I still think it would work. It's just an idea but thought I'd mention it.

File: WordPress/Classes/ViewRelated/Menus/MenusSelectionItemView.m (hunk 1 above)
Comment (etoledom) [reply WordPress/Classes/ViewRelated/Menus/MenusSelectionItemView.m]: Nice catch!

File: WordPress/Classes/ViewRelated/Menus/MenusSelectionView.m (hunk 2 above)
Comment (etoledom) [reply WordPress/Classes/ViewRelated/Menus/MenusSelectionView.m]: True that two `MenuSelectionitem`s to `MenusSelectionView` might be confusing, and it's used just on the `menus picker` use case.
Delegates are usually used to inform about stuff rather than requesting data, right? It would be better as a Data Source.

Since all the information is in the controller, I'll give it a try to set these labels in the controller itself, rather than passing the data to the views to set their own labels. Wdyt?

Comment (guarani) [reply WordPress/Classes/ViewRelated/Menus/MenusSelectionView.m]: > Delegates are usually used to inform about stuff rather than requesting data, right? It would be better as a Data Source.

I was just referring to the delegate pattern (I didn't make that clear), i.e. the use of a protocol to fetch stuff  — which covers "delegates" or "data sources".

> Since all the information is in the controller, I'll give it a try to set these labels in the controller itself, rather than passing the data to the views to set their own labels. Wdyt?

I think that would also work 👍

**Comment by etoledom** [comment]
@Guarani - Thanks for the review!

> I tested this and overall it works great, but I found one potential bug. Here's the steps to reproduce:
> 
> 1. In the top section (menu location), select a different selected menu location (changed from "Header" to "Social Menu" in my test)
> 2. With VoiceOver on, tap the header view (i.e. `MenusSelectionDetailView`) once to hear its accessibility message
> 
> I heard:
> 
> > Menu in area Header: No Menu. 2 menus available. Expanded. Button. [hint] Expands to select a different menu
> 
> but I expected:
> 
> > Menu in area **Social Menu**: No Menu. 2 menus available. Expanded. Button. [hint] Expands to select a different menu
> 
> because "Social Menu", not "Header", is the currently selected menu location.

Interesting, I couldn't reproduce the problem.
Each time I switch the Menu area, then the Menu picker updates and VoiceOver reads as expected.

This is what I do:
- Select and expand Menu Area picker
- Select a different menu area
- Select the Menu picker
  - Expect that VoiceOver reads the menu area name selected previously.


### About the labels:

#### Menu area picker
> Maybe we could shorten this to "3 menu areas"?

Cool let's do that!

#### Menu picker
> "Menu in Header area: 3 menus available. Button. [hint] Expands to select a different menu to edit."

I think it's important to mention the currently selected menu. I add the name of the area where this menu will be placed so the mental connection with the previous picker can be created. This UI can be a bit complex to understand even for non-voiceover users, so I'm trying to be extra careful with presenting the UI on a way that can be understood.

So I think it is important to mention the currently selected menu and in what menu area it is placed (even though the area is mentioned in the previous picker).

> Header menu area. 3 menu areas in this theme. Button. [hint] Expands to select a different menu area.

That reads better indeed, but I think it's important to give priority to what control it is, and second what information does it have. For example: This is the `Menu area` picker, and the selection is `Primary`.

If we start with the menu selected and then it changes, a VoiceOver user might think it is a different control all together.

**Comment by etoledom** [comment]
@Guarani - Ready for another 👀 🙏

**Comment by jkmassel** [comment]
We're cutting the `15.6` release today, so this PR is being bumped to `15.7`. If it needs to land in the `15.6` release, please merge it to the `release/15.6` branch, and DM me so that I can release a new beta build.

Thanks!

File: WordPress/Classes/ViewRelated/Menus/MenusSelectionItemView.m (hunk 1 above)
Comment (guarani) [reply WordPress/Classes/ViewRelated/Menus/MenusSelectionItemView.m]: Working great now, I hear:
- on selected items: "Footer menu. Selected. Button"
- on not selected items: "Primary. Button. Selects this item"
👍

File: WordPress/Classes/ViewRelated/Menus/MenuHeaderViewController.m [hunk 3]
```diff
@@ -139,4 +142,36 @@ - (void)selectionViewSelectedOptionForCreatingNewItem:(MenusSelectionView *)sele
...
+
+- (void)configureLocationAccessibility
+{
+    // Menu area: Header. 3 menu areas available. Button. [hint] Expands to select a different menu area.
+    NSString *format = NSLocalizedString(@"Menu area: %@, %d menu areas available", @"Screen reader string too choose a menu area to edit. %@ is the name of the menu area (Primary, Footer, etc...). %d is a number indicating the ammount of menu areas available.");
+    MenusSelectionItem *selectedItem = self.locationsView.selectedItem;
+    self.locationsView.accessibilityLabel = [NSString stringWithFormat:format,
+                                                        selectedItem.displayName,
+                                                        self.blog.menuLocations.count];//self.locationsView.subTitleLabel.text];
+    self.locationsView.accessibilityHint = NSLocalizedString(@"Expands to select a different menu area", @"Screen reader hint (non-imperative) about what does the site menu area selector button do.");
```
Comment (guarani) [inline WordPress/Classes/ViewRelated/Menus/MenuHeaderViewController.m:161]: From what I see here the accessibility hint is set irrespective of whether the item is selected or not, and I'm seeing some slightly inconsistent behavior. 
Here's the sequence of steps I do (with VoiceOver on):
1. Single tap the menu location button (`MenusSelectionDetailView`) to hear: the accessibility label plus the "Expands to select a different menu area" accessibility hint ✅
2. Double tap the same button to expand the options, now I hear: the accessibility label plus the accessibility value "Expanded" ✅
3. Single tap the same button (just to hear VoiceOver read it again), and I hear a different message: the accessibility label plus the "Expands to select a different menu area" accessibility hint ⚠️

That final message doesn't sound right, it should read the same message as in Step 2. Here's a video that includes sound (VoiceOver audio included): https://cloudup.com/cVFQgi9dJGV (31s)

**Review by guarani — APPROVED** [review]
I think this is ready, if you can't reproduce the new issue I mentioned in https://github.com/wordpress-mobile/WordPress-iOS/pull/14639/files#r475925103.

(It might be an iOS 14 beta thing.)

So if you can't reproduce this issue, feel free to merge as I think it's working great!

File: WordPress/Classes/ViewRelated/Menus/MenuHeaderViewController.m (hunk 3 above)
Comment (etoledom) [reply WordPress/Classes/ViewRelated/Menus/MenuHeaderViewController.m:161]: Thank you for noticing this!

For the hint to be the same is fine as long as the status shows that the control is already expanded. In this case, when I test the point 3, VoiceOver does say the value `Expanded` so I'd say it's Ok. I see in the video that this is also the case for you 👍 

After performing the action, is normal for VoiceOver to skip reading the Hint.

**Comment by etoledom** [comment]
Thank you @Guarani and @iamthomasbishop for the reviews! 🙏

File: WordPress/Classes/ViewRelated/Menus/MenuHeaderViewController.m (hunk 3 above)
Comment (guarani) [reply WordPress/Classes/ViewRelated/Menus/MenuHeaderViewController.m:161]: Ok thanks for clarifying 👍
